src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStep.java

```diff
@@ -97,26 +119,57 @@ public boolean start() throws Exception {
         @SuppressWarnings("rawtypes")
         @Override
         public OutputStream decorateLogger(Run build, final OutputStream logger) throws IOException, InterruptedException {
-            return new TeeOutputStream(logger, append(f));
+            return new TeeOutputStream(logger, stream = append(f, stream));
         }
 
         private static final long serialVersionUID = 1;
 
-    }
+        void close() throws IOException {
+            if (stream != null) {
+                stream.close();
+                stream = null;
+            }
+        }
 
-    /** @see FilePath#write() */
-    private static OutputStream append(FilePath fp) throws IOException, InterruptedException {
-        if (fp.getChannel() == null) {
-            File f = new File(fp.getRemote()).getAbsoluteFile();
-            if (!f.getParentFile().exists() && !f.getParentFile().mkdirs()) {
-                throw new IOException("Failed to create directory " + f.getParentFile());
+        private void writeObject(ObjectOutputStream oos) throws IOException, InterruptedException {
+            /*
+              This gets called either to serialize to allow resume or to transfer to a remote.
+              When resuming, the stream gets recreated as and when needed.
+              When transferring the stream needs to transfer as well.
+
+              The transferredToRemote is used to detect whether the stream got serialized as well
+              which is also guaranteeing backward compatibility when resuming after plug-in upgrade.
+             */
+            transferredToRemote = Channel.current() != null;
+            oos.defaultWriteObject();
```

Comment:
In the original PR I suggested that this be rewritten to use `writeReplace` for legibility and maintainability, but I do not consider it a blocker.